### PR TITLE
Adding mergeFields support

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This PR adds field mask support to set().

This is a port from https://github.com/firebase/firebase-js-sdk/pull/701

### CODE REVIEW

File: Firestore/Example/Tests/Integration/API/FIRDatabaseTests.mm [hunk 1]
```diff
@@ -256,6 +256,157 @@ - (void)testMergeReplacesArrays {
...
+         @"owner" : @{@"name" : @"Jonny", @"email" : @"abc@xyz.com"} };
+
+  NSDictionary<NSString *, id> *finalData = @{@"des" : @"Description", @"owner" : @"Sebastian"};
+
+  [self writeDocumentRef:doc data:initialData];
+
+  XCTestExpectation *completed =
+      [self expectationWithDescription:@"testCanSetASubsetOfFieldsUsingMask"];
+
+  [doc setData:@{@"desc" : @"NewDescription", @"owner" : @"Sebastian"}
```
Comment (gsoltis) [inline Firestore/Example/Tests/Integration/API/FIRDatabaseTests.mm]: The rest of the test uses `"des"`. They should probably all be `"desc"`?

Looks like all of the tests do this.

File: Firestore/Source/API/FSTUserDataConverter.mm [hunk 2]
```diff
@@ -424,9 +424,46 @@ - (FSTParsedSetData *)parsedMergeData:(id)input {
                                     path:absl::make_unique<FieldPath>(FieldPath::EmptyPath())];
   FSTObjectValue *updateData = (FSTObjectValue *)[self parseData:input context:context];
 
+  FieldMask convertedFieldMask;
+  std::vector<FieldTransform> convertedFieldTransform;
+
+  if (fieldMask) {
+    __block std::vector<FieldPath> fieldMaskPaths{};
+    [fieldMask
+        enumerateObjectsUsingBlock:^(id _Nonnull fieldPath, NSUInteger idx, BOOL *_Nonnull stop) {
```
Comment (gsoltis) [inline Firestore/Source/API/FSTUserDataConverter.mm]: I don't believe you need the `_Nonnull` annotations, as you are within an `NS_ASSUME_NONNULL` block.

File: Firestore/Example/Tests/Integration/API/FIRDatabaseTests.mm (hunk 1 above)
Comment (schmidt-sebastian) [reply Firestore/Example/Tests/Integration/API/FIRDatabaseTests.mm]: Update everything to "desc". Thanks for finding this.

File: Firestore/Source/API/FSTUserDataConverter.mm (hunk 2 above)
Comment (schmidt-sebastian) [reply Firestore/Source/API/FSTUserDataConverter.mm]: Done.

**Comment by gsoltis** [comment]
Please fix you whitespace to appear the linters though.

**Pull request merged**